Like feature and page ordering features

tameTNT opened the pull request:
**DO NOT MERGE BEFORE page-render functionality BRANCH**
This PR adds both a Like feature and, as a part of it, the ability to view the 'best' pages (i.e. by number of likes) or the ones that have been created most recently. New links have been added to the front page and some general redirects tweaked for usability improvements.

baru64 committed:
Merge pull request #7 from Artemis21/tameTNT-app-maker

tameTNT-app-maker

baru64 committed:
Add basic ModelForms

baru64 committed:
Add auth views and urls

Changes:
  - added basic auth, base and main page templates
  - added urls for auth and main page
  - added generic auth views and simple main view

Artemis21 committed:
Enforce PEP8

Artemis21 committed:
Merge pull request #15 from Artemis21/add-auth-views

Add auth views and urls

baru64 committed:
Add user view classes and templates

baru64 committed:
Add webpage and template views - work in progress

tameTNT committed:
Install imgkit

imgkit module installed via pipenv.
Pipfile and Pipfile.lock updated accordingly.

tameTNT committed:
Change use of os to Path module

Path module use is more pythonic and recommended by code jam organisers
over using os.join().

baru64 committed:
Fix static files

baru64 committed:
Add webpage and template views, urls and base templates

tameTNT committed:
Start create render function and set up template

README.md updated with new environment variable setup

tameTNT committed:
Merge remote-tracking branch 'origin/console-theme' into add-page-render-functionality

tameTNT committed:
Merge remote-tracking branch 'origin/artemis-dev' into add-page-render-functionality

tameTNT committed:
Merge remote-tracking branch 'origin/add-page_maker-app' into add-page-render-functionality

tameTNT committed:
Merge remote-tracking branch 'origin/add-base-views' into add-page-render-functionality

# Conflicts:
#	jolly-jellyfish/src/django_meta/urls.py
#	jolly-jellyfish/src/page_maker/models.py

tameTNT committed:
Simplify and expand README.md

Refine wording of Introduction and Idea sections.
Add more line breaks for readability when editing README.md file.
Add explanation of environment variable config to setup section.
Add complete language name to code section syntax highlighting options.

tameTNT committed:
Update models.py and adapt templates accordingly.

Change models.py (and be extension forms.py) with some path changes and
use of RelativeFilePathField (pipenv install required) instead of
models.FilePathField for easier stylesheet access within templates.

Reload the database files too -
delete original db and migrations and start afresh.
Delete residual file: generic_template.html no longer required.

Edit base.html to be able to access stylesheet by static path.
Edit webpage_list.html and webpage_view.html with minor formatting
changes for when there are no sites and to actually load user images +
prevent errors when no images present.

tameTNT committed:
Add sys-env  warning to render_img_thumbnail.py

Add try-except block to cover if user has not added WKHTML_TO_IMAGE to
system environment variables as specified in the readme.

tameTNT committed:
Merge remote-tracking branch 'origin/master' into add-page-render-functionality

baru64 committed:
Add comment views

Changes:
  - add comment section to webpage-detail
  - add comments to WebpageDetailView context
  - fix CommentCreate and WebpageCreate success url
  - add LOGIN_URL to settings

tameTNT committed:
Implement page thumbnail generation.

Thumbnail rendering function created as a custom templatetag.
This allows it to be executed from within a template when it is detected
that the page currently uses placeholder_img.png in its thumbnail field.

webpage_view.html also rearranged to make more logical sense.

tameTNT committed:
Simplify render_img_thumbnail.py tag functions

webpage_list.html adapted accordingly. Comments added to page code too.
More specificity added to models.py as well
to provide greater input freedom.
(typo corrected in views.py)

tameTNT committed:
Simplify render_img_thumbnail.py tag functions

webpage_list.html adapted accordingly. Comments added to page code too.
More specificity added to models.py as well
to provide greater input freedom.
(typo corrected in views.py)

tameTNT committed:
Merge remote-tracking branch 'origin/add-page-render-functionality' into add-page-render-functionality

# Conflicts:
#	jolly-jellyfish/src/page_maker/templates/page_maker/webpage_list.html

tameTNT committed:
Merge remote-tracking branch 'origin/master' into add-page-render-functionality

tameTNT committed:
Merge remote-tracking branch 'origin/add-base-views' into add-page-render-functionality

# Conflicts:
#	jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html
#	jolly-jellyfish/src/page_maker/views.py

tameTNT committed:
Improve and rework thumbnail image rendering

Thumbnails are now generated when the user reaches the page itself
(not the list page). This is accomplished within base.html.
This means thumbnails are generated immediately following page creation
(since user is directed to the actual webpage after filling in form).

Thumbnails are also automatically regenerated when a page is edited.

*However, as noted in the Discord chat, some underlying issues have been
discovered including background images (in css) not being loaded by
imgkit.
This is due to the underlying wkhtmltoimage binary used and hence
doesn't appear solvable using imgkit alone.
Any potential solutions - feel free to add.

Artemis21 committed:
Add global styling.

And extended version of the Windows 95 theme. Also added headers to pages as it selects the first element and styles it as a header.

Artemis21 committed:
Merge branch 'add-page-render-functionality' into add-global-style

Artemis21 committed:
Use Pathlib for paths for cross platform compability.

tameTNT committed:
Improve layout of 'best' and 'user' pages

Thumbnails and page links are now tabulated.
This makes the page more compact as well as easier to use.
(In prep. for standardised thumbnail image size
this make pages more uniform)

tameTNT committed:
Upgrade render_img_thumbnail.py to use Selenium

Selenium offers a major functional improvement over the previous imgkit.
Although it may be slightly slower (probs. due to browser load).

Update Pipfile and .lock with the new package requirements.
Update README.md with new download and env variable requirements.
Update models.py path gen function for Template.style_sheet due to newly
added site-wide style.css which should not be accessible
for user template creation.
Change affected stylesheet paths used in base.html

tameTNT committed:
Upgrade render_img_thumbnail.py to use Selenium

Selenium offers a major functional improvement over the previous imgkit.
Although it may be slightly slower (probs. due to browser load).

Update Pipfile and .lock with the new package requirements.
Update README.md with new download and env variable requirements.
Update models.py path gen function for Template.style_sheet due to newly
added site-wide style.css which should not be accessible
for user template creation.
Change affected stylesheet paths used in base.html

tameTNT committed:
Merge remote-tracking branch 'origin/add-page-render-functionality' into add-page-render-functionality

tameTNT committed:
Add Template preview support to thumbnail rendering

Add extended documentation and annotation to render_thumbnail.py.
Add new TemplateView to preview what a stylesheet looks like.

Selenium can now be run using either Firefox or Chromium.

tameTNT committed:
Update README.md with Firefox driver detail

Added readme information concerning use of the geckodriver for Firefox.

tameTNT committed:
Clarify template logic and fix url inconsistency

The new user_detail.html and webpage_list.html account for the default
value of a Webpage object - with 'placeholder_img' in the path -
not just a blanket check for the url (which will always be True).

For Template-object-related url templates,
one url used the pk while another used the templatename.
Both now use templatename.
(since this field is now unique in the Template model)

tameTNT committed:
Update jolly-jellyfish/README.md pipenv title

Clarity update in prep for potential future changes.

Co-authored-by: EXPLOSION <40616000+A5rocks@users.noreply.github.com>

tameTNT committed:
Update webpage_create.html with energy!

Bit more exciting this way!

Co-authored-by: EXPLOSION <40616000+A5rocks@users.noreply.github.com>

tameTNT committed:
Vastly improve readability of render_thumbnail.py

Thanks to @A5rocks for all the insight on the PR!
The file, esp. variable names, is now much more readable.
(and it also actually supports Firefox rather than just pretending to)

tameTNT committed:
Merge remote-tracking branch 'origin/add-page-render-functionality' into add-page-render-functionality

tameTNT committed:
Apply suggestions from code review - README.md

README.md changes for clarity - suggested on PR code review

tameTNT committed:
Apply suggestions from code review

Clarity and consistency improvements in models.py, urls.py and settings.py along with a correction in README.md

tameTNT committed:
Re-add missing comma

Trust me to break things when adding suggestions.

tameTNT committed:
Apply suggestions from code review - misc. html formatting

Polish HTML formatting including newlines at end-of-file and some typo and sylting corrections.

tameTNT committed:
Neaten HTML templates

Adjust indentation across templates to 4 spaces (in line with Python).
Remove functionless <div> wrappers.
Add newlines (inc. to EOF) and more spacing to cluttered files.

tameTNT committed:
Add Like model to models.py

Add new model to allow for like functionality where a user can like a
webpage to improve its ordering on the 'best pages' page.

tameTNT committed:
Add Like model to models.py

Add new model to allow for like functionality where a user can like a
webpage to improve its ordering on the 'best pages' page.

tameTNT committed:
Add like count/popularity column to 'best pages'

Add a new column to the webpage_list.html template which tracks the
total number of likes a Webpage has received from users.

tameTNT committed:
Add like count/popularity column to 'best pages'

Add a new column to the webpage_list.html template which tracks the
total number of likes a Webpage has received from users.

tameTNT committed:
Finish Like model and page ordering

Register Like model with admin.py.
Remove deprecated votes attribute from Webpage model - this is now
handled by the fk in the Like model.
Implement ordering in WebpageListView, so webpages are now sorted by
number of likes on the 'Best Pages' ('/pages/') page.

tameTNT committed:
Finish Like model and page ordering

Register Like model with admin.py.
Remove deprecated votes attribute from Webpage model - this is now
handled by the fk in the Like model.
Implement ordering in WebpageListView, so webpages are now sorted by
number of likes on the 'Best Pages' ('/pages/') page.

tameTNT committed:
Improve formatting and clarity

Remove redundant arguments in super() function
Add detail to comments

tameTNT committed:
Apply suggestions from code review - add trailing '/'

Improve consistency across views.py file

tameTNT committed:
Tidy TODOs and remove extraneous comments

Add new packages needed to requirements.txt
Add minor formatting changes to example_page.html

tameTNT committed:
Merge remote-tracking branch 'origin/like-feature-(and-page-ranking)' into like-feature-(and-page-ranking)

tameTNT committed:
Add more num likes labels and enable pagination

Edit urls.py to conform with no leading '/' convention.
Actually implement pagination code on webpage_list.html with buttons to
go to the next/previous page etc.

tameTNT committed:
Improve basic page functionality

Add '?next={{ request.path }}' to end of urls to ensure user is
redirected back to the page where they clicked the button. More of a
convenience feature than anything!

Added clarification to user_detail.html so that the page says if the
user hasn't created any pages.

Corrected typo in register.html.

tameTNT committed:
Fully add 'like' feature to relevant pages

Users can now, if logged in, like/unlike pages (but not their own!)
A webpage's like count is now displayed in various locations - including
on the 'best pages' page and the page itself (webpage_detail.html).

Minor sorting improvements in UserDetailView to sort by edit datetime.

tameTNT committed:
Add a 'recent pages' page

New url for recent pages, 'pages-recent/', and by extension for 'best'
also,'pages-top/'. These are both WebpageListView views, just with
different ordering values.
Update main.html to show this new navigation option.

Artemis21 committed:
Merge pull request #23 from Artemis21/use-platform-paths

Use Pathlib for paths for cross platform compability.

A5rocks commented:
I'll review after the rendering branch gets merged, just too much duplicated code I don't want to think about...

A5rocks committed:
Fix `Using pipenv` section

A5rocks committed:
Make code consistent

There were several things where this wasn't consistent, such as:
 - 4-space indentation for HTML, 2-space is convention (I think)
 - Indentation within if blocks (this is broken intentionally in the one place that remains)
 - Single-quotes instead of double-quotes
 - URLs shouldn't end with `/`

A5rocks commented:
I'm of the opinion we should close this PR in favor of https://github.com/Artemis21/summer-code-jam-2020/pull/29

A5rocks committed:
Merge branch 'add-page-render-functionality' into like-feature-(and-page-ranking)

A5rocks committed:
Make code more consistent

- Change HTML from 4-space indentation to 2-space indentation
- No indents in if blocks
- Prefer single-quotes unless the string has a contraction.
